File: 백준/Platinum/1948. 임계경로/임계경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static class Edge {
        int to;
        int time;

        Edge(int to, int time) {
            this.to = to;
            this.time = time;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        int[] indegree = new int[n + 1];

        List<List<Edge>> graph = new ArrayList<>();
        List<List<Edge>> reverseGraph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
            reverseGraph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken());
            int to = Integer.parseInt(st.nextToken());
            int time = Integer.parseInt(st.nextToken());

            graph.get(from).add(new Edge(to, time));
            reverseGraph.get(to).add(new Edge(from, time));
            indegree[to]++;
        }

        StringTokenizer st = new StringTokenizer(br.readLine());
        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        Queue<Integer> queue = new LinkedList<>();
        int[] maxTime = new int[n + 1];
        queue.offer(start);

        while (!queue.isEmpty()) {
            int cur = queue.poll();

            for (Edge edge : graph.get(cur)) {
                indegree[edge.to]--;
                maxTime[edge.to] = Math.max(maxTime[edge.to], maxTime[cur] + edge.time);

                if (indegree[edge.to] == 0) queue.offer(edge.to);
            }
        }

        int criticalEdgeCount = 0;
        boolean[] visited = new boolean[n + 1];
        queue.offer(end);
        visited[end] = true;

        while (!queue.isEmpty()) {
            int cur = queue.poll();

            for (Edge edge : reverseGraph.get(cur)) {
                if (maxTime[edge.to] + edge.time == maxTime[cur]) {
                    criticalEdgeCount++;

                    if (!visited[edge.to]) {
                        visited[edge.to] = true;
                        queue.offer(edge.to);
                    }
                }
            }
        }

        System.out.println(maxTime[end]);
        System.out.println(criticalEdgeCount);
    }
}

File: 백준/Platinum/1948. 임계경로/임계경로.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList
import java.util.Queue

data class Edge(val to: Int, val time: Int)

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val m = br.readLine().toInt()
    val indegree = IntArray(n + 1)

    val graph = List(n + 1) { mutableListOf<Edge>() }
    val reverseGraph = List(n + 1) { mutableListOf<Edge>() }

    repeat(m) {
        val (from, to, time) = br.readLine().split(" ").map { it.toInt() }
        graph[from].add(Edge(to, time))
        reverseGraph[to].add(Edge(from, time))
        indegree[to]++
    }

    val (start, end) = br.readLine().split(" ").map { it.toInt() }
    val queue: Queue<Int> = LinkedList()
    val maxTime = IntArray(n + 1)
    queue.offer(start)

    while (queue.isNotEmpty()) {
        val cur = queue.poll()

        for (edge in graph[cur]) {
            indegree[edge.to]--
            maxTime[edge.to] = maxOf(maxTime[edge.to], maxTime[cur] + edge.time)

            if (indegree[edge.to] == 0) queue.offer(edge.to)
        }
    }

    var criticalEdgeCount = 0
    val visited = BooleanArray(n + 1)
    queue.offer(end)
    visited[end] = true

    while (queue.isNotEmpty()) {
        val cur = queue.poll()

        for (edge in reverseGraph[cur]) {
            if (maxTime[edge.to] + edge.time == maxTime[cur]) {
                criticalEdgeCount++

                if (!visited[edge.to]) {
                    visited[edge.to] = true
                    queue.offer(edge.to)
                }
            }
        }
    }

    println(maxTime[end])
    println(criticalEdgeCount)
}
